myhpi/core/context.py

```diff
@@ -1,9 +1,36 @@
+from django.contrib.auth.models import Group
+from django.db.models import Q
 from wagtail.core.models import Page, Site
 
 
 def base_context(request):
     root_page = getattr(Site.find_for_request(request), "root_page", None)
+    pages = Page.objects.in_menu().child_of(root_page).specific()
+
+    # filter menu items based on which pages the user can view
+    def can_view(pages, usergroups):
+        page_list = []
+        for page in pages:
```

Comment:
- 1 page w/ 2 subpages w/ 2 subpages each: 17 Queries, 7ms
- 2 pages w/ 2 subpages each w/ 2 subpages each: 69 Queries, 11ms
- 4 pages w/ 2 subpages each w/ 2 subpages each: 109 Queries, 20ms
- 8 pages w/ 2 subpages each w/ 2 subpages each: 189 Queries, 27ms
Latencies have sometimes high fluctuations, so those values are not reliable. Django already does some query caching but idk if that will be enough in production.

Reply:
that sounds like a lot of queries :/

Reply:
I looked over the code with @jeriox. Right now, this function (which doesn't need to be defined in a function body and could be named `_get_menu_items`) fetches subpages in line 14 each call, and in case of leaf nodes evaluates the `visible_for.all()` queryset for that page. We should avoid evaluating querysets with each call.

We could fetch the whole tree of Pages and Menuitems at once and go through it "in python", without hitting the database using e.g. `get_descendants` on the root_page. I fear it wont be easy to come up with nice code handling the tree, but maybe the graph framework wagtail uses offers some tools for that.

For checking permissions, we first thought we could slap a `prefetch_related("visible_for")` on it, but Menuitems don't have that. Also, right now, there is no single base class defining `visible_for`, so we semantically need to go looking in a different db table for each Page Type. As discussed elsewhere, such a base class is to be introduced, though it might be annoying to have yet another table for that. (btw, why aren't you using djangos object permissions?)

As menu items still wont have `visible_for` you could fetch the "reverse": the list of all pages/page ids that are visible by the user (based on their group memberships). Then you can go walk the tree and check viewing permissions without needing extra queries.

Reply:
Base class has been added and process has been reworked. It now produces only 2 queries.